src/feature/hs/hs_client.c

```diff
@@ -1288,6 +1302,10 @@ client_desc_has_arrived(const smartlist_t *entry_conns)
     if (!hs_client_any_intro_points_usable(identity_pk, desc)) {
       log_info(LD_REND, "Hidden service descriptor is unusable. "
                         "Closing streams.");
+      /* Report the extended socks error code that we were unable to introduce
```

Comment:
I'm wondering if this is the right place to call this function. In particular, we seem to only get here when we get a new desc, but isn't that too late? This means that we fetched two descriptors before throwing the error page. Shouldn't we throw the error when we finish up all intros of the first descriptor? Perhaps we should be calling this from another place where `hs_client_any_intro_points_usable()` is used?

Reply:
Not sure... Tor client behavior has always been to try all intro points, if they all failed, refetch new descriptor and if it failed again, that means the service is gone and close SOCKS connection.

Thing is that the client doesn't know if the descriptor it has is out of date so the second retry is important for reachability but also attempt to get a newer descriptor before the cache entry expires.